### PR TITLE
Transition `Thread` to tailwind and refactor layout styling

### PR DESCRIPTION
**Pull request by lyzadanger**
This PR migrates the `Thread` component to Tailwind and attempts to untangle some complex styling to be more comprehensible and less fragile.

Minimal user-visible static changes but a small fix to a defect and a change to hover behavior in the left channel of replies (see below).

### User-visible changes

* Fixes a visual defect in which the leftmost edge of an annotation’s header got cut off by the toggle control for the thread. This was a result of the fiddly absolute positioning involved before these changes; this is fixed and should not occur in general again because the layout approach has been refactored.

  Before:

  1784BE13-DCF4-4663-AD89-AF400801DCAE

  After: 

  323A7B22-7B2D-4DC1-A792-35E0C9F8CCF3

* The left “channel” of a thread looks the same but the hover area is enlarged, such that mousing over that area highlights the thread’s extent more readily. This is easier shown by example and screen videos, see below.
* There is 1.5px more vertical space between each reply (from `6.5px` to `8px`). This gets the vertical layout rhythm on-scale.  This is _subtle_, but in the interest of trying to be exhaustive about calling out any change whatsoever, there you have it!

#### Left-channel hover

The left “channel” of a thread (except for top-level threads, which do not have a left channel because they cannot be collapsed) renders a toggle button with a directional caret and a vertical dashed line that shows the extent of the thread.
In static form these look identical after these changes.

Previously, this layout was implemented with a combination of absolute positioning, negative margins and a right border (dashed). The element that had the border-line was only 1px wide. There was a hover-darkening effect on the border-line and it was possible to hover over the line by positioning your mouse _just so_ or hovering the button:

thread-hovering-before

These changes remove some fiddly layout such that the left channel (flex item) naturally takes the width of the toggle button and the height of the thread's content. Now the hover area of the left channel is naturally larger:

thread-hovering-after

I’ve taken pains to make sure the mouse cursor is non-pointer, as the dashed-line is not clickable. There are a few ways we could go here:

* Land and keep as-is, with larger hover area but only toggle button is clickable. I do think the hover is helpful in that it highlights the extent of the associated thread, which can be hard to discern in deeply-nested threads otherwise.
* OTOH, maybe it’s confusing that the hovered thread-line can’t be interacted with. In that case, we could:
	* Make the hover-line clickable (collapse the associated thread), or
	* Only apply the hover-darkening effect on the thread-line when the toggle-button itself is hovered (a little more work but certainly doable)

Part of https://github.com/hypothesis/client/issues/4377

### CODE REVIEW

File: src/sidebar/components/Thread.js [hunk 1]
```diff
@@ -12,6 +12,91 @@ import ModerationBanner from './ModerationBanner';
...
+ * thread. When a thread is visible, an Annotation is rendered which in turn
+ * renders an AnnotationHeader. But when a thread is hidden, Annotation is not
+ * rendered: use this header instead for hidden top-level threads.
+ *
+ * @param {object} props
+ *   @param {Thread['annotation']} props.annotation
+ *   @param {number} props.replyCount
+ *   @param {boolean} props.threadIsCollapsed
+ */
+function HiddenThreadCardHeader({ annotation, ...restProps }) {
```
Comment (lyzadanger) [inline src/sidebar/components/Thread.js:28]: This sub-component existed, but I've renamed it and moved it top top of module as our conventions to this point have tended towards sub-components above default-exported component.

File: src/sidebar/components/Thread.js [hunk 2]
```diff
@@ -12,6 +12,91 @@ import ModerationBanner from './ModerationBanner';
...
+/**
+ * Render a "channel" to the left of a thread's content with a control for
+ * expanding/collapsing the thread and a visual vertical line showing the
+ * extent of the thread.
+ *
+ * @param {object} props
+ *   @param {boolean} props.threadIsCollapsed
+ *   @param {() => void} props.onToggleReplies
+ */
+function ThreadCollapseControl({ threadIsCollapsed, onToggleReplies }) {
```
Comment (lyzadanger) [inline src/sidebar/components/Thread.js:58]: This sub-component is new to these changes.

File: src/sidebar/components/Thread.js [hunk 3]
```diff
@@ -73,64 +158,64 @@ function Thread({ thread, threadsService }) {
...
+            className={classnames(
+              // Pull this list to the left to bring it closer to the left edge
+              // of the ThreadCard and give more space for nested replies' content
+              '-ml-3'
+            )}
+            data-testid="thread-children"
+          >
             {visibleChildren.map(child => (
-              <li key={child.id}>
+              <li key={child.id} className="mt-2">
```
Comment (lyzadanger) [inline src/sidebar/components/Thread.js:218]: This top margin was previously set on the thread container element, but only for replies. So it seemed more appropriate for each reply-containing `li` to set its top margin instead. This is what defines spacing between the annotations in a thread.

File: src/sidebar/components/test/Thread-test.js [hunk 4]
```diff
@@ -138,12 +140,6 @@ describe('Thread', () => {
       assert.calledOnce(fakeStore.setExpanded);
       assert.calledWith(fakeStore.setExpanded, replyThread.id, false);
     });
-
-    it('assigns an appropriate CSS class to the element', () => {
```
Comment (lyzadanger) [inline src/sidebar/components/test/Thread-test.js:142]: Neither this nor the other test removed tested for anything particularly meaningful, IMHO.

File: src/styles/sidebar/components/_index.scss [hunk 5]
```diff
@@ -24,17 +24,42 @@
...
+      180deg,
+      var(--tw-gradient-from),
+      var(--tw-gradient-from) 4px,
+      transparent 4px,
+      transparent 6px,
+      var(--tw-gradient-from) 6px
+    );
+    background-size: 1px 100%;
+  }
+}
```
Comment (lyzadanger) [inline src/styles/sidebar/components/_index.scss:62]: I spent a fair amount of time trying to determine the right abstraction for this and I've timed out for the moment. I'm certain this will move, and almost certain it will be renamed, but I'm not sure yet to where and to what. The current `.bg-` prefix on the classname echoes Tailwind naming conventions for anything to do with backgrounds. I'm not advocating for it, just explaining.

This is the first time I've surfaced what is under the covers for Tailwind by directly referencing some CSS properties. Lots of flexibility here.

**Comment by chrisshaw** [comment]
tl;dr I concur with your option 1: land as is, keep it unclickable.

<details>
<summary>A few more thoughts</summary>

I did a little digging with some tools I thought may have similar use cases to ours and ultimately ended up with 3 actual analog: Reddit, Workflowy, and LogSeq. (Many tools either solve this problem differently or not at all.)

Reddit has you click this line to collapse the thread. This seems to still garner help requests on how to actually collapse the thread. Having the little caret would likely help in our case, however, as they don't.

Workflowy leaves it unclickable.

LogSeq does something interesting, where clicking it collapses it **to that level**. So, instead of collapsing all the comments, it collapses only those deeper than direct, level 1 comments. Just toying with it, I did see the utility, though not a lot, and even after I knew how it worked it will still confusing what happened when I clicked that line.

I will say personally my first instinct said, "If I click this, it will jump me to the top of this thread." Which I quite like, actually, because from there I could then click the collapse button. But I wouldn't ship this without testing with users first, as I can imagine that could be a lot of motion.

P.S. sorry I took so long to comment. You shouldn't feel the need to wait if I don't hop on this as I'd originally intended.

</details>

**Comment by robertknight** [comment]
> OTOH, maybe it’s confusing that the hovered thread-line can’t be interacted with

My immediate thought is that it would be convenient if the thread line could be clicked to expand/collapse the thread, as it is a much easier target to hit than the small caret box. However I don't think we need to decide that before trying to land this - this PR is an improvement as-is.

**Comment by robertknight** [comment]
OK, this behavior looks broken and it applies both in production and on this branch:

Full thread visible:

All replies visible

Search for text that appears only in the first reply:

Filter active

Note that the "Show 2 more in conversation" button appears twice.

File: src/sidebar/components/Thread.js [hunk 6]
```diff
@@ -73,64 +158,64 @@ function Thread({ thread, threadsService }) {
...
+      >
         {annotationContent}
 
-        {showHiddenToggle && (
-          <>
+        {isHidden && (
+          <div className="space-y-2">
             {!thread.parent && (
-              <ThreadHeader
+              <HiddenThreadCardHeader
```
Comment (robertknight) [inline src/sidebar/components/Thread.js:192]: If you move the `isEditing` logic into `AnnotationHeader` then you could skip the helper and just render `thread.parent && annotation && <AnnotationHeader .../>` here.

**Review by robertknight — APPROVED** [review]
LGTM, except that I think the `isHidden` variable name is confusing. I also had a suggestion that would enable eliminating the wrapper component around `AnnotationHeader`. That might be better done in a follow-up PR.
